src/main/java/org/gaul/s3proxy/S3ProxyHandler.java

```diff
@@ -547,6 +559,17 @@ public final void doHandle(HttpServletRequest baseRequest,
             }
         }
 
+        // Validate container name
+        if (!uri.equals("/") && !isValidContainer(path[1])) {
+            if (method.equals("PUT") &&
+                    (path.length <= 2 || path[2].isEmpty()) &&
+                    !("".equals(request.getParameter("acl"))))  {
+                throw new S3Exception(S3ErrorCode.INVALID_BUCKET_NAME);
+            } else {
+                throw new S3Exception(S3ErrorCode.NO_SUCH_BUCKET);
```

Comment:
The underlying `BlobStore` operations should throw `ContainerNotFoundException`.  Unfortunately the jclouds does not propagate this.  Could you look at [JCLOUDS-193](https://issues.apache.org/jira/browse/JCLOUDS-193)?